flocker/control/test/test_persistence.py

```diff
@@ -249,3 +333,202 @@ def test_complex_keys(self):
+    """
+    v1_config = json.dumps({"version": 1})
+
+    def test_error_on_undefined_migration_path(self):
+        """
+        A ``MissingMigrationError`` is raised if a migration path
+        from one version to another cannot be found in the supplied
+        migration class.
+        """
+        e = self.assertRaises(
+            MissingMigrationError,
+            migrate_configuration, 1, 4, self.v1_config, StubMigration
+        )
+        expected_error = (
+            u'Unable to find a migration path for a version 3 to '
+            u'version 4 configuration. No migration method '
+            u'upgrade_from_v3 could be found.'
+        )
+        self.assertEqual(e.message, expected_error)
+
+    def test_sequential_migrations(self):
+        """
+        A migration from one configuration version to another will
+        sequentially perform all necessary upgrades, e.g. v1 to v2 followed
+        by v2 to v3.
+        """
+        # Get a valid v2 config.
+        v2_config = migrate_configuration(1, 2, self.v1_config, StubMigration)
+        # Perform two sequential migrations to get from v1 to v3, starting
+        # with a v1 config.
+        result = migrate_configuration(1, 3, self.v1_config, StubMigration)
+        # Compare the v1 --> v3 upgrade to the direct result of the
+        # v2 --> v3 upgrade on the v2 config, Both should be identical
+        # and valid v3 configs.
+        self.assertEqual(result, StubMigration.upgrade_from_v2(v2_config))
+
+
+UUIDS = st.basic(generate=lambda r, _: UUID(int=r.getrandbits(128)))
+
+DATASETS = st.builds(Dataset, dataset_id=UUIDS, maximum_size=st.integers())
+
+# Constrain primary to be True so that we don't get invariant errors from Node
+# due to having two differing manifestations of the same dataset id.
+MANIFESTATIONS = st.builds(
+    Manifestation, primary=st.just(True), dataset=DATASETS)
+SIMPLE_TEXT = st.text(
+    alphabet=string.letters, min_size=4, max_size=20, average_size=12
+)
+IMAGES = st.builds(DockerImage, tag=SIMPLE_TEXT, repository=SIMPLE_TEXT)
+NONE_OR_INT = st.one_of(
+    st.just(None),
+    st.integers()
+)
+ST_PORTS = st.integers(min_value=1, max_value=65535)
+PORTS = st.builds(
+    Port,
+    internal_port=ST_PORTS,
+    external_port=ST_PORTS
+)
+LINKS = st.builds(
+    Link,
+    local_port=ST_PORTS,
+    remote_port=ST_PORTS,
+    alias=SIMPLE_TEXT
+)
+APPLICATIONS = st.builds(
+    Application, name=SIMPLE_TEXT, image=IMAGES,
+    ports=st.sets(PORTS, max_size=10),
+    links=st.sets(LINKS, max_size=10),
+    environment=st.dictionaries(keys=SIMPLE_TEXT, values=SIMPLE_TEXT),
+    memory_limit=NONE_OR_INT,
+    cpu_shares=NONE_OR_INT,
+    running=st.booleans()
+)
+FILEPATHS = st.text(alphabet=string.printable).map(FilePath)
+VOLUMES = st.builds(
+    AttachedVolume, manifestation=MANIFESTATIONS, mountpoint=FILEPATHS)
+APPLICATIONS_WITH_VOLUMES = st.tuples(
+    APPLICATIONS, VOLUMES).map(lambda (a, v): a.set(volume=v))
```

Comment:
There's no need for a separate `APPLICATIONS_WITH_VOLUMES`. Instead, the `APPLICATIONS` strategy should set `volume` to `st.none() | VOLUMES`. That way, we'll test all cases.

This will require some changes further down, as `_build_node` assumes that `app.volume` is not `None`.